### ISSUE TITLE
Python bindings documentation

### ISSUE THREAD

**Issue by wvxvw** [SimpleITK/SimpleITK#1437]
I was only able to find PDF version here: https://readthedocs.org/projects/simpleitk/downloads/pdf/master/

This is... less than ideal. It prevents me from being able to refer to your classes from documentation in my project. Also, PDF is just an all-around awful format for documentation in terms of searching, indexing etc.

Is there any reason you aren't publishing HTML documentation?

**Comment by wvxvw** [comment]
Also, I couldn't find `index.xml` for the Doxygen documentation (which is otherwise impossible to link to automatically because of the garbage in generated link names). Is there any chance you will include that in published documentation?

**Comment by gdevenyi** [comment]
https://simpleitk.readthedocs.io ?

**Comment by blowekamp** [comment]
I'll look into adding generation of Doxygen XML.

**Comment by wvxvw** [comment]
@gdevenyi It doesn't have documentation for the code. It has examples, guides etc. I'm looking for "developer documentation", a.k.a. "API documentation". Something that gives *authoritative* and *exhaustive* list of all programming objects available to the library user.

I.e. I'm not interested in learning how to use the library. I need to be able to refer the reader of my documentation to the relevant object in your documentation whenever my code uses your library.

**Comment by gdevenyi** [comment]
I think this is what you're looking for.

https://simpleitk.org/doxygen/v2_0/html/namespaceitk_1_1simple.html

**Comment by wvxvw** [comment]
@gdevenyi No. That's C++ documentation. I want Python documentation.

**Comment by gdevenyi** [comment]
As far as I understand, the python bindings are automatically generated from the C++ ones, and hence the documentation is the same

P.S. "latest" link instead, https://simpleitk.org/doxygen/latest/html/

**Comment by dave3d** [comment]
The PDF that you've linked to is the same content as the ReadTheDocs site.  Neither of them have a full Python API documentation.  You pretty much have to read the C++ docs and translate it into Python in your head.

**Comment by wvxvw** [comment]
@gdevenyi Well, C++ documentation is generated automatically too, and C++ template code is generated automatically too... why should this make any difference / why not generate Python documentation automatically too?

@dave3d There are multiple problems with that. Suppose I could do that, why would you expect every Python user to be able to do that? They came to use a Python library, but now they need to read the documentation for the language they aren't using / most likely even don't understand?

Secondly, C++ documentation has a lot of warts that are only relevant to C++ code. Who in Python world cares if you have a non-default copy constructor? How to tell if `std::string` will be a `bytes` object in Python or an `str`, or maybe it has a buffer interface, or maybe it's exposed as a NumPy array?

The module names are all wrong. For example, `SimpleITK.ImageFileReader` is the Python fully qualified name for `itk::simple::ImageFileReader`... maybe. I'm guessing here, but since there aren't too many classes with that name, I'm pretty sure about that, though I might be wrong.

Finally, Doxigen generated documentation is impossible to reference automatically, unless there's the `index.xml`. Even with `index.xml` it's a lot of manual labor and research to figure out what Python entities map to what C++ entities. It would give me the way to find the proper links to C++ entities if I can guess them, but I'm not sure I'll be able to guess them in every case.

**Comment by dave3d** [comment]
I would never argue that this is an ideal, or even good solution.  But with limited man power, the task of re-writing the entire ITK API documentation for Python seems like an incredibly daunting task.  As always, we welcome any and all contributions from the community.

**Comment by blowekamp** [comment]
We can generate the Doxygen XML with the current processes. If that is useful to you please let us know and I can look into it.

Much of the class documentation in SimpleITK comes from ITK, and then it is copied to Python too. It can contain very useful information about the algorithm, but some times it includes ITK implementation details. In the SimpleITK C++ Doxygen documentation this can make some sense and is nicely linked back to ITK Doxygen documentation for further details. As Python doc strings it is not so good, but the information can be helpful so it's there as part of the Python library.

It is possible for Sphinx to generate documentation from the Python package, but requires the library to already be built which takes some times, and would cause some dependency issue between RtD and the CI production of binaries. If someone looks into getting Sphinx to produce reasonable Python docs, and the community things they are more helpful thank confusing we could look into getting RtD to use the nightly builds or something.

**Comment by wvxvw** [comment]
> We can generate the Doxygen XML with the current processes. If that is useful to you please let us know

Yes, it's better than what's there right now. I'd appreciate it.

I'm not a great fan of Sphinx. I'm using it because it's integrated with the rest of Python documentation. Recently, I had to become more familiar with the source of `autodoc` in particular. I might try to research it further to understand if it'd be possible to generate anything useful from the docstrings already in SimpleITK modules, though I wouldn't hold your breath. It's hard to get sensible results with pure Python...

**Comment by blowekamp** [comment]
> It's hard to get sensible results with pure Python...

If that has been your experience then we will have quite a bit more difficulty with the SWIG wrapping.

One possibility is to just select certain classes and methods where the documentation is generated.

**Comment by wvxvw** [comment]
You might not like the sound of it... :) But, I was thinking about the way forward as taking the `index.xml` and using some XSLT "exciting" code to transform it into something that Docutils understands.

A long, long time ago I had to do something similar to ActionScript documentation (its generation was based on XSL transformations). Though I'd need to see what I'm up against. I'm not really an XSL "wizard".

**Comment by CsatiZoltan** [comment]
@wvxvw 
> Finally, Doxigen generated documentation is impossible to reference automatically

It is possible with Sphinx. I was looking for a way to do this, when I came across with the following setting in the `conf.py` file:
https://github.com/drcandacemakedamoore/cleanX/blob/main/docs/conf.py#L75-L80
This allows you to write a docstring like this:
```
Attributes
----------
    image : :sitk:`Image`
        Description
```
which will create a hyperlink to the [corresponding C++ documentation](https://simpleitk.org/doxygen/latest/html/classitk_1_1simple_1_1Image.html).

If you want to link a general documentation page on the RTD site, use Interpshinx as usual:
https://github.com/drcandacemakedamoore/cleanX/blob/9e29304a26577abf1b7894eb3050177e392f5832/docs/conf.py#L106
In the docstring, you can then write e.g.
```
See Also
--------
    :doc:`simpleitk:link_DicomSeriesReader_docs
```
bringing you to the [documentation](https://simpleitk.readthedocs.io/en/master/link_DicomSeriesReader_docs.html).

Here is a screenshot from the Sphinx documentation of my project, demonstrating what I wrote above:

![Sphinx](https://user-images.githubusercontent.com/12742890/138091186-c4633590-8abb-436a-aa89-a01c939989a5.png)

**Comment by blowekamp** [comment]
Thank you for your patience. We have just added a Github Actions workflow name "Doxygen" which generates a "doxygen-xml" artifact.

The processes for publishing the artifact for releases, and nightly bill will be developed shortly.

**Comment by blowekamp** [comment]
The Doxygen XML has been included in the release since 2.2.0
